basic/product-marketplace/src/lib.rs

```diff
@@ -0,0 +1,393 @@
+                                        id : (*product).id, 
+                                        name : (*product).name.clone(),
+                                        price : (*product).price,
+                                }); 
+                    products.push(product.to_string()); 
+                }
+            }
+            
+            info!("products : {}", products.join(";"));
+
+            return result;
+        }
+
+        pub fn buy_product(&mut self, product_id : u128, 
+                                      city : String, 
+                                      street : String, 
+                                      zip_code : String, 
+                                      payment : Bucket) -> (Bucket ,Bucket)
+        {
+            assert!(payment.resource_address() == self.token_type , "token address must match");
+            assert!(self.products_for_sale.contains_key(&product_id), "product not found");
+            assert!(self.product_is_available_for_sale(product_id), "product is not available");
+            let product = self.products_for_sale.get_mut(&product_id).unwrap(); 
+            let total_amount : Decimal = self.buy_fees + (*product).price; 
+            assert!(payment.amount() >= total_amount, "payment amount must be greather than or equal {}", total_amount); 
+
+            let buyer_address : PostalAddress = PostalAddress { city : city, 
+                                                                street : street, 
+                                                                zip_code : zip_code
+                                                               } ;   
+
+            let buyer_nft_id = self.buyer_nft_id_by_product_id.get(&product_id).unwrap(); 
+            let buyer_badge = self.buyer_minter_badge_vault.authorize(|auth| {
+                return self.buyer_badge_def.mint_nft(*buyer_nft_id, BuyerNftData {
+                     product_id : product_id,
+                     fees : self.buy_fees,
+                     has_been_sent_by_seller : false 
+                }, auth)
+            }); 
+
+           
+            // set buyer address
+            let seller_nft_id : u128 = *self.seller_nft_id_by_product_id.get(&product_id).unwrap();
+            let mut seller_nft_data  : SellerNftData = self.seller_product_badge_def.get_nft_data(seller_nft_id);  
+            
+            seller_nft_data.buyer_address = buyer_address; 
+            seller_nft_data.product_has_been_purchase = true; 
+            self.seller_product_minter_badge_vault.authorize(|a|  {
+                self.seller_product_badge_def.update_nft_data(seller_nft_id, seller_nft_data,a); 
+            }); 
+
+             // take fees 
+            self.fees_vault.put(payment.take(self.buy_fees));
+            // hold payments
+            let vault = self.payment_by_buyer_nft_id.entry(*buyer_nft_id).or_insert(Vault::new(self.token_type));
+            vault.put(payment.take((*product).price));
```

Comment:
You can do this in a single line with the `.and_modify` method:
```
self.payment_by_buyer_nft_id.entry(*buyer_nft_id)
                .and_modify(|vault| vault.put(payment.take(product.price)))
                .or_insert(Vault::with_bucket(payment.take(product.price)));
```

Reply:
Ok